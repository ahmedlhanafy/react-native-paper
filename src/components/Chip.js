/* @flow */

import * as React from 'react';
import {
  View,
  StyleSheet,
  Animated,
  TouchableWithoutFeedback,
} from 'react-native';
import color from 'color';
import Icon from './Icon';
import Surface from './Surface';
import Text from './Typography/Text';
import TouchableRipple from './TouchableRipple';
import withTheme from '../core/withTheme';
import type { Theme } from '../types';
import type { IconSource } from './Icon';

const AnimatedSurface = Animated.createAnimatedComponent(Surface);

type Props = {
  /**
   * Mode of the chip.
   * - `flat` - flat chip without outline
   * - `outlined` - chip with an outline
   */
  mode?: 'flat' | 'outlined',
  /**
   * Text content of the `Chip`.
   */
  children: React.Node,
  /**
   * Icon to display for the `Chip`. Both icon and avatar cannot be specified.
   */
  icon?: IconSource,
  /**
   * Avatar to display for the `Chip`. Both icon and avatar cannot be specified.
   */
  avatar?: React.Node,
  /**
   * Display the chip as selected.
   */
  selected?: boolean,
  /**
   * Disables the chip. `onPress` function won't execute.
   */
  disabled?: boolean,
  /**
   * Function to execute on press.
   */
  onPress?: () => mixed,
  /**
   * Function to execute on delete. The delete button appears only when this prop is specified.
   */
  onDelete?: () => mixed,
  style?: any,
  /**
   * @optional
   */
  theme: Theme,
};

type State = {
  elevation: Animated.Value,
};

/**
 * Chips can be used to display entities in small blocks.
 *
 * <div class="screenshots">
 *   <img src="screenshots/chip-1.png" />
 *   <img src="screenshots/chip-2.png" />
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { Chip } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <Chip icon="info" onPress={() => {}}>Example Chip</Chip>
 * );
 * ```
 */
class Chip extends React.Component<Props, State> {
  static defaultProps = {
    mode: 'flat',
    disabled: false,
    selected: false,
    style: {},
  };

  state = {
    elevation: new Animated.Value(0),
  };

  _handlePressIn = () => {
    Animated.timing(this.state.elevation, {
      toValue: 4,
      duration: 200,
    }).start();
  };

  _handlePressOut = () => {
    Animated.timing(this.state.elevation, {
      toValue: 0,
      duration: 150,
    }).start();
  };

  render() {
    const {
      mode,
      children,
      icon,
      avatar,
      selected,
      disabled,
      onPress,
      onDelete,
      style,
      theme,
    } = this.props;
    const { dark, colors } = theme;

    const backgroundColor =
      mode === 'outlined'
        ? colors.background
        : color(colors.text)
            .alpha(disabled ? 0.05 : 0.12)
            .rgb()
            .string();
    const textColor = disabled
      ? colors.disabled
      : color(colors.text)
          .alpha(dark ? 0.7 : 0.87)
          .rgb()
          .string();
    const iconColor = disabled
      ? colors.disabled
      : color(colors.text)
          .alpha(dark ? 0.7 : 0.54)
          .rgb()
          .string();
    const selectedColor = color(colors.text)
      .alpha(mode === 'outlined' ? 0.1 : 0.3)
      .rgb()
      .string();

    return (
      <AnimatedSurface
        style={[
          styles.touchable,
          {
            elevation: this.state.elevation,
          },
          style,
        ]}
      >
        <TouchableRipple
          style={styles.touchable}
          onPress={onPress}
          onPressIn={this._handlePressIn}
          onPressOut={this._handlePressOut}
          underlayColor={selectedColor}
          disabled={disabled}
        >
          <View
            style={[
              styles.content,
              {
                backgroundColor,
                borderColor: mode === 'outlined' ? colors.text : 'transparent',
              },
              selected
                ? {
                    backgroundColor: selectedColor,
                  }
                : null,
            ]}
          >
            {avatar && !icon
              ? /* $FlowFixMe */
                React.cloneElement(avatar, {
                  /* $FlowFixMe */
                  style: [styles.avatar, avatar.props.style],
                })
              : null}
            {icon || selected ? (
              <View
                style={[styles.icon, avatar ? styles.avatarSelected : null]}
              >
                {/* $FlowFixMe */}
                <Icon
                  source={selected ? 'done' : icon}
                  color={avatar ? colors.background : iconColor}
                  size={18}
                />
              </View>
            ) : null}
            <Text
              numberOfLines={1}
              style={[
                styles.text,
                {
                  color: textColor,
                  marginRight: onDelete ? 4 : 8,
                  marginLeft: avatar || icon || selected ? 4 : 8,
                },
              ]}
            >
              {children}
            </Text>
            {onDelete ? (
              <TouchableWithoutFeedback onPress={onDelete}>
                <View style={styles.delete}>
                  <Icon source="cancel" size={16} color={iconColor} />
                </View>
              </TouchableWithoutFeedback>
            ) : null}
          </View>
        </TouchableRipple>
      </AnimatedSurface>
    );
  }
}

const styles = StyleSheet.create({
  touchable: {
    borderRadius: 16,
  },
  content: {
    borderRadius: 16,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-around',
    paddingHorizontal: 4,
    borderWidth: StyleSheet.hairlineWidth,
  },
  icon: {
    paddingHorizontal: 4,
    paddingVertical: 7,
  },
  delete: {
    paddingHorizontal: 4,
    paddingVertical: 8,
  },
  text: {
    marginVertical: 8,
  },
  avatar: {
    width: 24,
    height: 24,
    borderRadius: 12,
    marginRight: 4,
  },
  avatarSelected: {
    position: 'relative',
    left: -30,
    marginRight: -30,
  },
});

export default withTheme(Chip);
