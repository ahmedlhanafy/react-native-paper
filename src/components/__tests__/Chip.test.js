/* @flow */

import * as React from 'react';
import renderer from 'react-test-renderer';
import Chip from '../Chip';

it('renders chip with onPress', () => {
  const tree = renderer
    .create(<Chip text="Example Chip" onPress={() => {}} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders chip with icon', () => {
  const tree = renderer
    .create(<Chip text="Example Chip" icon="info" />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders deletable chip', () => {
  const tree = renderer
    .create(<Chip text="Example Chip" icon="info" onDelete={() => {}} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
