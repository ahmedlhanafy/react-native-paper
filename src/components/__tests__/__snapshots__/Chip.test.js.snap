// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders chip with icon 1`] = `
<View
  collapsable={undefined}
  style={
    Array [
      Object {
        "backgroundColor": "#ffffff",
      },
      Object {
        "backgroundColor": "#ffffff",
      },
      0,
      Object {
        "borderRadius": 16,
        "elevation": 0,
      },
    ]
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "borderRadius": 16,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 16,
            "borderWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-around",
            "paddingHorizontal": 4,
          },
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.12)",
            "borderColor": "transparent",
          },
          null,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "paddingHorizontal": 4,
              "paddingVertical": 7,
            },
            null,
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          pointerEvents="none"
          style={
            Array [
              Object {
                "color": "rgba(0, 0, 0, 0.54)",
                "fontSize": 18,
              },
              Object {
                "backgroundColor": "transparent",
              },
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "Helvetica Neue",
            },
            Array [
              Object {
                "marginVertical": 8,
              },
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "marginLeft": 4,
                "marginRight": 8,
              },
            ],
          ]
        }
        theme={
          Object {
            "colors": Object {
              "accent": "#03dac4",
              "background": "#f6f6f6",
              "disabled": "rgba(0, 0, 0, 0.26)",
              "error": "#B00020",
              "placeholder": "rgba(0, 0, 0, 0.38)",
              "primary": "#6200ee",
              "surface": "#ffffff",
              "text": "#000000",
            },
            "dark": false,
            "fonts": Object {
              "light": "HelveticaNeue-Light",
              "medium": "HelveticaNeue-Medium",
              "regular": "Helvetica Neue",
              "thin": "HelveticaNeue-Thin",
            },
            "roundness": 4,
          }
        }
      >
        Example Chip
      </Text>
    </View>
  </View>
</View>
`;

exports[`renders chip with onPress 1`] = `
<View
  collapsable={undefined}
  style={
    Array [
      Object {
        "backgroundColor": "#ffffff",
      },
      Object {
        "backgroundColor": "#ffffff",
      },
      0,
      Object {
        "borderRadius": 16,
        "elevation": 0,
      },
    ]
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "borderRadius": 16,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 16,
            "borderWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-around",
            "paddingHorizontal": 4,
          },
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.12)",
            "borderColor": "transparent",
          },
          null,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "Helvetica Neue",
            },
            Array [
              Object {
                "marginVertical": 8,
              },
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "marginLeft": 8,
                "marginRight": 8,
              },
            ],
          ]
        }
        theme={
          Object {
            "colors": Object {
              "accent": "#03dac4",
              "background": "#f6f6f6",
              "disabled": "rgba(0, 0, 0, 0.26)",
              "error": "#B00020",
              "placeholder": "rgba(0, 0, 0, 0.38)",
              "primary": "#6200ee",
              "surface": "#ffffff",
              "text": "#000000",
            },
            "dark": false,
            "fonts": Object {
              "light": "HelveticaNeue-Light",
              "medium": "HelveticaNeue-Medium",
              "regular": "Helvetica Neue",
              "thin": "HelveticaNeue-Thin",
            },
            "roundness": 4,
          }
        }
      >
        Example Chip
      </Text>
    </View>
  </View>
</View>
`;

exports[`renders deletable chip 1`] = `
<View
  collapsable={undefined}
  style={
    Array [
      Object {
        "backgroundColor": "#ffffff",
      },
      Object {
        "backgroundColor": "#ffffff",
      },
      0,
      Object {
        "borderRadius": 16,
        "elevation": 0,
      },
    ]
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "borderRadius": 16,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 16,
            "borderWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-around",
            "paddingHorizontal": 4,
          },
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.12)",
            "borderColor": "transparent",
          },
          null,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "paddingHorizontal": 4,
              "paddingVertical": 7,
            },
            null,
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          pointerEvents="none"
          style={
            Array [
              Object {
                "color": "rgba(0, 0, 0, 0.54)",
                "fontSize": 18,
              },
              Object {
                "backgroundColor": "transparent",
              },
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "Helvetica Neue",
            },
            Array [
              Object {
                "marginVertical": 8,
              },
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "marginLeft": 4,
                "marginRight": 4,
              },
            ],
          ]
        }
        theme={
          Object {
            "colors": Object {
              "accent": "#03dac4",
              "background": "#f6f6f6",
              "disabled": "rgba(0, 0, 0, 0.26)",
              "error": "#B00020",
              "placeholder": "rgba(0, 0, 0, 0.38)",
              "primary": "#6200ee",
              "surface": "#ffffff",
              "text": "#000000",
            },
            "dark": false,
            "fonts": Object {
              "light": "HelveticaNeue-Light",
              "medium": "HelveticaNeue-Medium",
              "regular": "Helvetica Neue",
              "thin": "HelveticaNeue-Thin",
            },
            "roundness": 4,
          }
        }
      >
        Example Chip
      </Text>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "paddingHorizontal": 4,
            "paddingVertical": 8,
          }
        }
        testID={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          pointerEvents="none"
          style={
            Array [
              Object {
                "color": "rgba(0, 0, 0, 0.54)",
                "fontSize": 16,
              },
              Object {
                "backgroundColor": "transparent",
              },
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders outlined disabled chip 1`] = `
<View
  collapsable={undefined}
  style={
    Array [
      Object {
        "backgroundColor": "#ffffff",
      },
      Object {
        "backgroundColor": "#ffffff",
      },
      0,
      Object {
        "borderRadius": 16,
        "elevation": 0,
      },
    ]
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "borderRadius": 16,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 16,
            "borderWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-around",
            "paddingHorizontal": 4,
          },
          Object {
            "backgroundColor": "#f6f6f6",
            "borderColor": "#000000",
          },
          null,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "Helvetica Neue",
            },
            Array [
              Object {
                "marginVertical": 8,
              },
              Object {
                "color": "rgba(0, 0, 0, 0.26)",
                "marginLeft": 8,
                "marginRight": 8,
              },
            ],
          ]
        }
        theme={
          Object {
            "colors": Object {
              "accent": "#03dac4",
              "background": "#f6f6f6",
              "disabled": "rgba(0, 0, 0, 0.26)",
              "error": "#B00020",
              "placeholder": "rgba(0, 0, 0, 0.38)",
              "primary": "#6200ee",
              "surface": "#ffffff",
              "text": "#000000",
            },
            "dark": false,
            "fonts": Object {
              "light": "HelveticaNeue-Light",
              "medium": "HelveticaNeue-Medium",
              "regular": "Helvetica Neue",
              "thin": "HelveticaNeue-Thin",
            },
            "roundness": 4,
          }
        }
      >
        Example Chip
      </Text>
    </View>
  </View>
</View>
`;

exports[`renders selected chip 1`] = `
<View
  collapsable={undefined}
  style={
    Array [
      Object {
        "backgroundColor": "#ffffff",
      },
      Object {
        "backgroundColor": "#ffffff",
      },
      0,
      Object {
        "borderRadius": 16,
        "elevation": 0,
      },
    ]
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "borderRadius": 16,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 16,
            "borderWidth": 0.5,
            "flexDirection": "row",
            "justifyContent": "space-around",
            "paddingHorizontal": 4,
          },
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.12)",
            "borderColor": "transparent",
          },
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.3)",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "paddingHorizontal": 4,
              "paddingVertical": 7,
            },
            null,
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          pointerEvents="none"
          style={
            Array [
              Object {
                "color": "rgba(0, 0, 0, 0.54)",
                "fontSize": 18,
              },
              Object {
                "backgroundColor": "transparent",
              },
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "Helvetica Neue",
            },
            Array [
              Object {
                "marginVertical": 8,
              },
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "marginLeft": 4,
                "marginRight": 8,
              },
            ],
          ]
        }
        theme={
          Object {
            "colors": Object {
              "accent": "#03dac4",
              "background": "#f6f6f6",
              "disabled": "rgba(0, 0, 0, 0.26)",
              "error": "#B00020",
              "placeholder": "rgba(0, 0, 0, 0.38)",
              "primary": "#6200ee",
              "surface": "#ffffff",
              "text": "#000000",
            },
            "dark": false,
            "fonts": Object {
              "light": "HelveticaNeue-Light",
              "medium": "HelveticaNeue-Medium",
              "regular": "Helvetica Neue",
              "thin": "HelveticaNeue-Thin",
            },
            "roundness": 4,
          }
        }
      >
        Example Chip
      </Text>
    </View>
  </View>
</View>
`;
